### PR TITLE
Adding Rule Ids for ABS v2 (from ASB v1) via MC reason codes

### PR DESCRIPTION
**Pull request by MariusNi**
## Description

Adding Rule Ids for ABS v2 (from ASB v1) via MC reason codes. Also fixing couple of security issues:

- Some temporary files created with unrestricted access could be compromised by attackers. Added access restriction.
- Reading from system files such as ones under /proc that do not have size set but contain data needs a refactoring of an old common function since IoT Hub times in order to work with all files, including these special ones.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [x] I added unit-tests to validate my changes. All unit tests are passing.
- [x] I have merged the latest `main` branch prior to this PR submission.
- [x] I submitted this PR against the `main` branch.

### CODE REVIEW

File: src/common/commonutils/FileUtils.c [hunk 1]
```diff
@@ -38,6 +37,22 @@ char* LoadStringFromFile(const char* fileName, bool stopAtEol, void* log)
...
+                        {
+                            string = temp;
+                            memset(&string[i], 0, currentSize - i);
+                        }
+                        else
+                        {
+                            FREE_MEMORY(string);
+                            break;
+                        }
+                    }
```
Comment (jepio) [inline src/common/commonutils/FileUtils.c:55]: is there any test that exercises this new code path (ie. reads more than 1024 bytes)?

File: src/common/tests/CommonUtilsUT.cpp [hunk 2]
```diff
@@ -94,6 +94,19 @@ TEST_F(CommonUtilsTest, LoadStringFromSingleByteFile)
     EXPECT_TRUE(Cleanup(m_path));
 }
 
+TEST_F(CommonUtilsTest, LoadStringFromZeroLengthFile)
```
Comment (jepio) [inline src/common/tests/CommonUtilsUT.cpp:97]: Zero length files worked even before this change. We need a new test case that specifically targets "zero length but non-empty files" (like procfs).

File: src/common/commonutils/FileUtils.c [hunk 3]
```diff
@@ -20,15 +22,12 @@ char* LoadStringFromFile(const char* fileName, bool stopAtEol, void* log)
     {
         if (LockFile(file, log))
         {
-            fseek(file, 0, SEEK_END);
-            fileSize = ftell(file);
-            fseek(file, 0, SEEK_SET);
-
-            string = (char*)malloc(fileSize + 1);
-            if (string)
+            if (NULL != (string = (char*)malloc(initialSize)))
```
Comment (jepio) [inline src/common/commonutils/FileUtils.c:25]: The `(char *)` cast is present here but not on line 45 with the `realloc` call. I'd say drop it here for consistency (it's not necessary).

**Review by jepio — COMMENTED** [review]
Can you separate the RuleId change from the bugfixes, they are not related to this change.

Can you cleanup the commits - group related changes, add descriptive titles and descriptions - it helps people trying to get familiar with the project.

This changes look significant - won't this trigger a need to perform DI re-validation?

File: src/common/commonutils/FileUtils.c [hunk 4]
```diff
@@ -58,6 +73,8 @@ static bool SaveToFile(const char* fileName, const char* mode, const char* paylo
 
     if (fileName && mode && payload && (0 < payloadSizeBytes))
     {
+        RestrictFileAccessToCurrentAccountOnly(fileName);
```
Comment (jepio) [inline src/common/commonutils/FileUtils.c:76]: Is `SaveToFile` only used for existing files, or can the file be created by the `fopen` call that happens later? Looking at it from a logical point of view:
- if the file already exists you're changing the permissions (who says they should be restricted?)
- if the file doesn't exist then the `Restrict` call happens too early and will silently fail

File: src/common/commonutils/FileUtils.c [hunk 5]
```diff
@@ -932,6 +949,8 @@ int ReplaceMarkedLinesInFile(const char* fileName, const char* marker, const cha
         {
             if (NULL != (tempHandle = fopen(tempFileName, "w")))
             {
+                RestrictFileAccessToCurrentAccountOnly(tempFileName);
```
Comment (jepio) [inline src/common/commonutils/FileUtils.c:952]: This is racy - the file is first created with potentially wide open permissions and later restricted, so there is a window where it can be manipulated.

File: src/common/commonutils/FileUtils.c (hunk 1 above)
Comment (MariusNi) [reply src/common/commonutils/FileUtils.c:55]: I've tested in the real end to end scenarios with the Watchers, the Agent, the Platform, and the Security Baseline etc with both 1024 and very short buffer size, suboptimal for run time, but good for spotting regressions, of just 5 bytes, for example. I'm considering adding a new test case that tests specifically for 1024+ bytes buffers. Thanks for the feedback.

Comment (MariusNi) [reply src/common/commonutils/FileUtils.c:55]: There will be more PRs following up that will add more test coverage.

File: src/common/tests/CommonUtilsUT.cpp (hunk 2 above)
Comment (MariusNi) [reply src/common/tests/CommonUtilsUT.cpp:97]: Once this will be merged, more PRs will follow to expand the test coverage. For this test case that you suggested in particular, please feel free to propose one via a new PR. unittests are not functional tests and crafting such a file of 0 length but which contains data needs to be done in the test itself, not rely on profcs, unless it is something that works on all target distros.

Comment (MariusNi) [reply src/common/tests/CommonUtilsUT.cpp:97]: Thanks for your feedback!

File: src/common/commonutils/FileUtils.c (hunk 3 above)
Comment (MariusNi) [reply src/common/commonutils/FileUtils.c:25]: Well spotted, thanks! Will do

File: src/common/commonutils/FileUtils.c (hunk 1 above)
Comment (MariusNi) [reply src/common/commonutils/FileUtils.c]: Added here, see CommonUtilsTest!LoadStringFromBigLengthFile.

File: src/common/commonutils/FileUtils.c (hunk 4 above)
Comment (MariusNi) [reply src/common/commonutils/FileUtils.c:76]: Only files that already exist. One core principle for our code is that all files that we write to are restricted for access. The system files that we write to as part of the security baseline have their access permissions set separately.

File: src/common/commonutils/FileUtils.c (hunk 5 above)
Comment (MariusNi) [reply src/common/commonutils/FileUtils.c:952]: Please see last iteration. The file is now emptied after being access restricted, before writing to it.

**Comment by MariusNi** [comment]
> Can you separate the RuleId change from the bugfixes, they are not related to this change.
> 
> Can you cleanup the commits - group related changes, add descriptive titles and descriptions - it helps people trying to get familiar with the project.
> 
> The changes look significant - won't this trigger a need to perform DI re-validation?

As discussed, these fixes are taken together and are validated together with the refactor for rule ids. Yes, we will have a new validation and signoff on this. Let's continue this conversation there.

File: src/common/commonutils/FileUtils.c [hunk 6]
```diff
@@ -5,13 +5,15 @@
...
+    int currentSize = 0;
     FILE* file = NULL;
-    int fileSize = 0;
     int i = 0;
     int next = 0;
     char* string = NULL;
+    char* temp = NULL;
 
-    if ((NULL == fileName) || (-1 == access(fileName, F_OK)))
+    if (false == FileExists(fileName))
```
Comment (robertschaedler3) [inline src/common/commonutils/FileUtils.c:16]: nit: `if (!FileExists(fileName))`

**Review by robertschaedler3 — APPROVED** [review]

File: src/common/commonutils/FileUtils.c (hunk 6 above)
Comment (MariusNi) [reply src/common/commonutils/FileUtils.c:16]: Returns bool and if you look at other instances in this file, same explicit comparison is used.

File: src/adapters/mc/OsConfigResource.c [hunk 7]
```diff
@@ -484,7 +490,13 @@ void MI_CALL OsConfigResource_Invoke_GetTargetResource(
     MI_UNREFERENCED_PARAMETER(instanceName);
 
     const char* auditPassed = "Audit passed";
-    const char* auditFailed = "Audit failed. See /var/log/osconfig*";
+    const char* auditFailed = "Audit failed. See See /var/log/osconfig_nrp.*";
```
Comment (AhmedBM) [inline src/adapters/mc/OsConfigResource.c:493]: Duplicated "See"

Comment (MariusNi) [reply src/adapters/mc/OsConfigResource.c:493]: Good catch! Will fix.

Comment (MariusNi) [reply src/adapters/mc/OsConfigResource.c]: Done, please see last iteration

**Review by AhmedBM — APPROVED** [review]
👌

File: src/adapters/mc/asb/AzureLinuxBaseline.mof [hunk 8]
```diff
@@ -1222,6 +1301,7 @@ instance of OsConfigResource as $OsConfigResource78ref
 instance of OsConfigResource as $OsConfigResource79ref
 {
     ResourceID = "Ensure all bootloaders have password protection enabled";
```
Comment (simonjaeger) [inline src/adapters/mc/asb/AzureLinuxBaseline.mof:1303]: Is it OK that the ResourceID is not mapping exactly with the corresponding Name ("All bootloaders should have password protection enabled" in this instance)? Is that agreed?

File: src/adapters/mc/asb/AzureLinuxBaseline.mof [hunk 9]
```diff
@@ -1838,7 +1958,8 @@ instance of OsConfigResource as $OsConfigResource119ref
 
 instance of OsConfigResource as $OsConfigResource120ref
 {
-    ResourceID = "Ensure the isdnutils-base package is not installed";
+    ResourceID = "Ensure the isdn service is disabled and the isdnutils-base package is not installed";
+    ruleId = "51ebf409-911a-4d92-9d3a-1e331e7e4b27";
```
Comment (simonjaeger) [inline src/adapters/mc/asb/AzureLinuxBaseline.mof:1962]: nit: `ruleId` - should be PascalCase?

File: src/adapters/mc/ssh/LinuxSshServerSecurityBaseline.mof [hunk 10]
```diff
@@ -305,6 +324,7 @@ instance of OsConfigResource as $OsConfigResource18ref
 instance of OsConfigResource as $OsConfigResource19ref
 {
     ResourceID = "Ensure that the SSH port is configured";
+    RuleId = "1310042D-193F-414A-A7E3-B2C4E89AA537";
```
Comment (simonjaeger) [inline src/adapters/mc/ssh/LinuxSshServerSecurityBaseline.mof:327]: Just curious. Is this a new `RuleId`? It looks different from others (also uppercase).

File: src/common/asb/Asb.c [hunk 11]
```diff
@@ -629,6 +629,222 @@ static char* g_desiredEnsureDefaultDenyFirewallPolicyIsSet = NULL;
...
+    { "Ensure the auditd package is installed", "f9fd03d2-75e4-4564-84a9-4e955f1e7c30", "EnsureAuditdInstalled" },
+    { "Ensure the auditd service is running", "c146c4a4-5eb6-4205-88da-5a71a82f2d45", "EnsureAuditdServiceIsRunning" },
+    { "Ensure the SNMP Server is not enabled", "ca1aea32-3969-49ab-abfc-2c5796a9a8bb", "EnsureSnmpServerIsDisabled" },
+    { "Ensure the rsync service is not enabled", "63aba603-b1f8-40df-82c5-38915452ce23", "EnsureRsynServiceIsDisabled" },
+    { "Ensure the NIS server is not enabled", "b4ad3fdd-7b68-4b11-a3ed-84b37a68b995", "EnsureNisServerIsDisabled" },
+    { "Ensure the rsh client is not installed", "6d441f31-f888-4f4f-b1da-7cfc26263e3f", "EnsureRshClientNotInstalled" },
+    { "Ensure SMB V1 with Samba is disabled", "7624efb0-3026-4c72-8920-48d5be78a50e", "EnsureSmbWithSambaIsDisabled" }
+};
+
+int AsbIsValidResourceIdRuleId(const char* resourceId, const char* ruleId, const char* payloadKey, void* log)
```
Comment (simonjaeger) [inline src/common/asb/Asb.c:811]: Is this meant to reject any `RuleId` which is not known? Is there a concern with hard-coding that?

File: src/common/asb/Asb.c [hunk 12]
```diff
@@ -629,6 +629,222 @@ static char* g_desiredEnsureDefaultDenyFirewallPolicyIsSet = NULL;
...
+    for (i = 0; i < numRules; i++)
+    {
+        if (0 == strcmp(payloadKey, (g_rules[i]).payloadKey))
+        {
+            if ((NULL != resourceId) && (0 != strcmp(resourceId, (g_rules[i]).resourceId)))
+            {
+                OsConfigLogError(log, "AsbIsValidRuleIdAndName: resourceId for rule '%s' of '%s' (instead of '%s') is invalid", payloadKey, resourceId, (g_rules[i]).resourceId);
+                result = ENOENT;
+            }
+            else if ((NULL != ruleId) && (0 != strcmp(ruleId, (g_rules[i]).ruleId)))
```
Comment (simonjaeger) [inline src/common/asb/Asb.c:832]: Is it possible to use `strncmp` here as `RuleId` length is known?

File: src/tests/universal-nrp-e2e/UniversalNRP.Tests.ps1 [hunk 13]
```diff
@@ -32,13 +32,11 @@ Describe 'Validate Universal NRP' {
 
         It 'Ensure resons are properly populated' {
```
Comment (simonjaeger) [inline src/tests/universal-nrp-e2e/UniversalNRP.Tests.ps1:33]: Should there also be a test to make sure that `RuleId` is included? Seems like the below test is checking for `BaselineSettingCompliant` and `BaselineSettingNotCompliant` only.

**Review by simonjaeger — APPROVED** [review]

File: src/tests/universal-nrp-e2e/UniversalNRP.Tests.ps1 (hunk 13 above)
Comment (MariusNi) [reply src/tests/universal-nrp-e2e/UniversalNRP.Tests.ps1:33]: The actual test in in the new function in Asb.c, AsbIsValidResourceIdRuleId. If a rule id or rule name is changed from expected, the reason code will state that and fail this test in this PS script.  The NRP itself validates that the input from MOF from MC and AzP is correct and that is invoked for this test. It works quite well actually.

Comment (MariusNi) [reply src/tests/universal-nrp-e2e/UniversalNRP.Tests.ps1:33]: Please reactivate if you still have questions. Thanks

File: src/common/asb/Asb.c (hunk 11 above)
Comment (MariusNi) [reply src/common/asb/Asb.c:811]: Any rule id and rule name that does not match the expected values for that rule. We need the reference, and the choice here is to put this into the code so the NRP can also use it at run time. The audits will still work, but the reason codes will  signal the problems. Please reactivate this if you still have questions.

File: src/common/asb/Asb.c (hunk 12 above)
Comment (MariusNi) [reply src/common/asb/Asb.c:832]: Good catch. Will do change.

File: src/adapters/mc/ssh/LinuxSshServerSecurityBaseline.mof (hunk 10 above)
Comment (MariusNi) [reply src/adapters/mc/ssh/LinuxSshServerSecurityBaseline.mof:327]: The new rule ids are all in clean all uppercase characters. There are total of 6 such rules here. This is one subtle way to tell which ones are, for example.

Comment (MariusNi) [reply src/adapters/mc/ssh/LinuxSshServerSecurityBaseline.mof:327]: Please reactivate if you still have questions or comments

File: src/adapters/mc/asb/AzureLinuxBaseline.mof (hunk 9 above)
Comment (MariusNi) [reply src/adapters/mc/asb/AzureLinuxBaseline.mof:1962]: Good catch! This works as is but yes, it should be PascalCase, this is a typo. Will fix.

File: src/adapters/mc/asb/AzureLinuxBaseline.mof (hunk 8 above)
Comment (MariusNi) [reply src/adapters/mc/asb/AzureLinuxBaseline.mof:1303]: All names in here follow a similar consistent pattern. There are some slight changes from v1 etc. Names will be one last time reviewed before release.

File: src/adapters/mc/asb/AzureLinuxBaseline.mof (hunk 9 above)
Comment (MariusNi) [reply src/adapters/mc/asb/AzureLinuxBaseline.mof]: Fixed. Please see last iteration. Thanks for the feedback!

File: src/common/asb/Asb.c (hunk 12 above)
Comment (MariusNi) [reply src/common/asb/Asb.c]: Fixed. Please see last iteration. Thanks for the feedback!

**Review by simonjaeger — APPROVED** [review]

**Review by simonjaeger — APPROVED** [review]
